pkg/config/spec.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"encoding/json"
+	"io"
+	"math"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/shirou/gopsutil/mem"
+	"k8s.io/klog/v2"
+)
+
+func rename(name string) string {
+	replacements := []struct {
+		old, new string
+	}{
+		{"(R)", ""},
+		{"(r)", ""},
+		{"CPU", ""},
+		{"Processor", ""},
+		{"processor", ""},
+		{"@", ""},
+		{"®", ""},
+		{",", ""},
+	}
+
+	for _, r := range replacements {
+		name = strings.ReplaceAll(name, r.old, r.new)
+	}
+
+	// Remove "n-Bit Multi-Core" pattern
+	re := regexp.MustCompile(`\d+-Bit Multi-Core`)
+	name = re.ReplaceAllString(name, "")
+
+	// Remove anything in parentheses or square brackets
+	re = regexp.MustCompile(`[\(\[].*?[\)\]]`)
+	name = re.ReplaceAllString(name, "")
+
+	// Remove frequency pattern
+	re = regexp.MustCompile(`\d+(\.\d+)?\s?[GMgm][Hh]z`)
+	name = re.ReplaceAllString(name, "")
+
+	return strings.TrimSpace(name)
+}
+
+func formatProcessor(processor string) string {
+	name := rename(processor)
+	name = strings.ToLower(name)
+	// replace spaces and hyphens with an underscore
+	name = strings.ReplaceAll(name, "-", "_")
+	fields := strings.Fields(name)
+	name = strings.Join(fields, "_")
+	return strings.ReplaceAll(name, "_v", "v")
+}
+
+func formatVendor(vendor string) string {
+	vendor = strings.ToLower(vendor)
+	parts := strings.Fields(vendor)
+	if len(parts) > 0 {
+		vendor = parts[0]
+	}
+	vendor = strings.ReplaceAll(vendor, "-", "_")
+	vendor = strings.ReplaceAll(vendor, ",", "")
+	vendor = strings.ReplaceAll(vendor, "'", "")
+	return vendor
+}
+
+func roundToNearestHundred(value float64) int {
+	return int(math.Round(value/100) * 100)
+}
+
+func generateSpec() *MachineSpec {
+	spec := &MachineSpec{}
+	spec.Cores = -1
+	spec.Memory = -1
+	spec.Frequency = -1
+	spec.Chips = 1
```

Comment:
this field is not set. the example in config map shows `chips: 2`

Reply:
Updated.